tests/PasswordHashingProvider.cpp

```diff
@@ -0,0 +1,185 @@
+    
+    ASSERT_EQ(salt1.length(), expectedSaltLength);
+}
+
+TEST_F(PasswordHashingProviderTests, CanGenerateDifferentSaltsDuringFirstUseFromTwoInstances)
+{
+    PasswordHashingProvider h1(defaultSaltLength, defaultHashLength);
+    PasswordHashingProvider h2(defaultSaltLength, defaultHashLength);
+    
+    std::string salt1 = h1.generateSalt();
+    std::string salt2 = h2.generateSalt();
+
+    ASSERT_NE(salt1, salt2);
+}
+
+TEST_F(PasswordHashingProviderTests, CanGenerateDifferentSaltsFromOneInstance)
+{
+    PasswordHashingProvider h1(defaultSaltLength, defaultHashLength);
+        
+    std::string salt1 = h1.generateSalt();
+    std::string salt2 = h1.generateSalt();
+
+    ASSERT_NE(salt1, salt2);
+}
+
+TEST_F(PasswordHashingProviderTests, WhenUsingDefaultConstructorHashLengthIsEqualTo16)
+{
+    const unsigned givenHashLength = 16;
+    const unsigned expectedHashLength = 2*givenHashLength;
+    PasswordHashingProvider h1 = PasswordHashingProvider();
+        
+    std::string hash1 = h1.generateHash(defaultSalt, defaultPassword);
+    
+    ASSERT_EQ(hash1.length(), expectedHashLength);
+}
+
+TEST_F(PasswordHashingProviderTests, CanGenerateHashWithGivenLength)
+{
+    const unsigned givenHashLength = 32;
+    const unsigned expectedHashLength = 2*givenHashLength;
+    PasswordHashingProvider h1(defaultSaltLength, givenHashLength);
+        
+    std::string hash1 = h1.generateHash(defaultSalt, defaultPassword);
+    
+    ASSERT_EQ(hash1.length(), expectedHashLength);
+}
+
+TEST_F(PasswordHashingProviderTests, CannotGenerateHashShorterThan8)
+{
+    const unsigned givenHashLength = 4;
+    const unsigned expectedHashLength = 2*8;
+    PasswordHashingProvider h1(defaultSaltLength, givenHashLength);
+        
+    std::string hash1 = h1.generateHash(defaultSalt, defaultPassword);
+    
+    ASSERT_EQ(hash1.length(), expectedHashLength);
+}
+
+TEST_F(PasswordHashingProviderTests, CannotGenerateHashLongerThan64)
+{
+    const unsigned givenHashLength = 128;
+    const unsigned expectedHashLength = 2*64;
+    PasswordHashingProvider h1(defaultSaltLength, givenHashLength);
+        
+    std::string hash1 = h1.generateHash(defaultSalt, defaultPassword);
+    
+    ASSERT_EQ(hash1.length(), expectedHashLength);
+}
+
+TEST_F(PasswordHashingProviderTests, CanConstructClassInstanceWithCopyConstructor)
+{
+    const unsigned givenHashLength = 64;
+    PasswordHashingProvider h1(defaultSaltLength, givenHashLength);
+    
+    PasswordHashingProvider h2(h1);
+    std::string hash1 = h1.generateHash(defaultSalt, defaultPassword);
+    std::string hash2 = h2.generateHash(defaultSalt, defaultPassword);
+    
+    ASSERT_EQ(hash1.length(), hash2.length());
+    ASSERT_EQ(hash1, hash2);
+}
+
+TEST_F(PasswordHashingProviderTests, CanConstructClassInstanceWithCopyAssignmentOperator)
+{
+    const unsigned givenHashLength = 64;
+    PasswordHashingProvider h1(defaultSaltLength, givenHashLength);
+    
+    PasswordHashingProvider h2 = h1;
```

Comment:
this construction doesn't use copy assignment operator. It uses copy constructor.
Copy assigment is used when you have already initialized object and you are assigning a new value to it like here:
```cpp
PasswordHashingProvider h1(defaultSaltLength, givenHashLength);
PasswordHashingProvider h2(defaultSaltLength, givenHashLength);
h2 = h1;
```

Reply:
True! Already corrected.